synapse/visibility.py

```diff
@@ -94,12 +106,16 @@ async def filter_events_for_client(
                 [event.event_id for event in events],
             )
 
-    types = (_HISTORY_VIS_KEY, (EventTypes.Member, user_id))
+    non_outlier_event_ids = frozenset(
+        e.event_id for e in events if not e.internal_metadata.outlier
+    )
 
-    # we exclude outliers at this point, and then handle them separately later
-    event_id_to_state = await storage.state.get_state_for_events(
-        frozenset(e.event_id for e in events if not e.internal_metadata.outlier),
-        state_filter=StateFilter.from_types(types),
+    # Grab the history visibility and membership for each of the events. That's all we
+    # need to know in order to filter them.
+    event_id_to_state = await storage.state._get_state_for_events_when_filtering_for_client(
+        # we exclude outliers at this point, and then handle them separately later
+        event_ids=non_outlier_event_ids,
+        user_id_viewing_events=user_id,
```

Comment:
Since we've changed to more individual queries, it's hard to just test all of the queries from a real life transaction against the database and be 100% confident that everything is better. Can we test this against `matrix.org` somehow to verify it's an improvement before merging (deploy to a special worker)?